drivers/serial/uart_esp32.c

```diff
@@ -1,58 +1,541 @@
+		data->irq_cb(data->irq_cb_data);
+	}
+}
+
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+static const struct uart_driver_api uart_esp32_api = {
+	.poll_in = uart_esp32_poll_in,
+	.poll_out = uart_esp32_poll_out,
+	.err_check = uart_esp32_err_check,
+	.configure =  uart_esp32_configure,
+	.config_get = uart_esp32_config_get,
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	.fifo_fill = uart_esp32_fifo_fill,
+	.fifo_read = uart_esp32_fifo_read,
+	.irq_tx_enable = uart_esp32_irq_tx_enable,
+	.irq_tx_disable = uart_esp32_irq_tx_disable,
+	.irq_tx_ready = uart_esp32_irq_tx_ready,
+	.irq_rx_enable = uart_esp32_irq_rx_enable,
+	.irq_rx_disable = uart_esp32_irq_rx_disable,
+	.irq_tx_complete = uart_esp32_irq_tx_complete,
+	.irq_rx_ready = uart_esp32_irq_rx_ready,
+	.irq_err_enable = uart_esp32_irq_err_enable,
+	.irq_err_disable = uart_esp32_irq_err_disable,
+	.irq_is_pending = uart_esp32_irq_is_pending,
+	.irq_update = uart_esp32_irq_update,
+	.irq_callback_set = uart_esp32_irq_callback_set,
+#endif  /* CONFIG_UART_INTERRUPT_DRIVEN */
 };
 
-DEVICE_AND_API_INIT(esp32_uart, "ROMUART",
-		    esp32_uart_init, NULL, NULL,
-		    PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
-		    &esp32_uart_api);
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+#define ESP32_UART_IRQ_HANDLER_DECL(idx) \
+	static void uart_esp32_irq_config_func_##idx(struct device *dev)
+
+#define ESP32_UART_IRQ_HANDLER_FUNC(idx) \
+	.irq_config_func = uart_esp32_irq_config_func_##idx,
+
+#define ESP32_UART_IRQ_HANDLER(idx)					     \
+	static void uart_esp32_irq_config_func_##idx(struct device *dev)     \
+	{								     \
+		esp32_rom_intr_matrix_set(0, ETS_UART##idx##_INTR_SOURCE,    \
+					  DT_UART_ESP32_PORT_##idx##_IRQ_0); \
+		IRQ_CONNECT(DT_UART_ESP32_PORT_##idx##_IRQ_0,		     \
+			    1,						     \
+			    uart_esp32_isr,				     \
+			    DEVICE_GET(uart_esp32_##idx),		     \
+			    0);						     \
+		irq_enable(DT_UART_ESP32_PORT_##idx##_IRQ_0);		     \
+	}
+#else
+#define ESP32_UART_IRQ_HANDLER_DECL(idx)
+#define ESP32_UART_IRQ_HANDLER_FUNC(idx)
+#define ESP32_UART_IRQ_HANDLER(idx)
+
+#endif
+
+#define ESP32_UART_INIT(idx)								\
+ESP32_UART_IRQ_HANDLER_DECL(idx);						\
+static const struct uart_esp32_config uart_esp32_cfg_port_##idx = {		\
+	.dev_conf = {								\
+		.base = (u8_t *)DT_INST_##idx##_XTENSA_ESP32_UART_BASE_ADDRESS,	\
+		.sys_clk_freq = CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC,		\
+		ESP32_UART_IRQ_HANDLER_FUNC(idx)				\
+	},									\
+										\
+	.peripheral = {								\
+		.clk = DPORT_UART##idx##_CLK_EN,				\
+		.rst = DPORT_UART##idx##_RST,					\
+	},									\
+										\
+	.signals = {								\
+		.tx_out = U##idx##TXD_OUT_IDX,					\
+		.rx_in = U##idx##RXD_IN_IDX,					\
+		.rts_out = U##idx##RTS_OUT_IDX,					\
+		.cts_in = U##idx##CTS_IN_IDX,					\
+	},									\
+										\
+	.pins = {								\
+		.tx = DT_INST_##idx##_XTENSA_ESP32_UART_TX_PIN,			\
+		.rx = DT_INST_##idx##_XTENSA_ESP32_UART_RX_PIN,			\
+		.rts = DT_INST_##idx##_XTENSA_ESP32_UART_RTS_PIN,		\
```

Comment:
Setting RTS and CTS should be optional as it is in DT. If they are not defined in boards DT, building uart driver fails.
```
[ 91%] Building C object zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/uart_esp32.c.obj
zephyr/drivers/serial/uart_esp32.c:500:10: error: 'DT_INST_0_XTENSA_ESP32_UART_RTS_PIN' undeclared here (not in a function); did you mean 'DT_INST_0_XTENSA_ESP32_UART_RX_PIN'?
   .rts = DT_INST_##idx##_XTENSA_ESP32_UART_RTS_PIN,  \
          ^~~~~~~~
zephyr/drivers/serial/uart_esp32.c:532:1: note: in expansion of macro 'ESP32_UART_INIT'
 ESP32_UART_INIT(0);
 ^~~~~~~~~~~~~~~
zephyr/drivers/serial/uart_esp32.c:501:10: error: 'DT_INST_0_XTENSA_ESP32_UART_CTS_PIN' undeclared here (not in a function); did you mean 'DT_INST_0_XTENSA_ESP32_UART_TX_PIN'?
   .cts = DT_INST_##idx##_XTENSA_ESP32_UART_CTS_PIN,  \
          ^~~~~~~~
zephyr/drivers/serial/uart_esp32.c:532:1: note: in expansion of macro 'ESP32_UART_INIT'
 ESP32_UART_INIT(0);
 ^~~~~~~~~~~~~~~
make[2]: *** [zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/build.make:63: zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/uart_esp32.c.obj] Error 1
make[1]: *** [CMakeFiles/Makefile2:1497: zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
```

Reply:
Oh, I forgot about that !
Nice catch.
I will add also UART_X_FLOW_CONTROL to Kconfig to control the feature, dunno if there is a way through dts

Reply:
Linux uses `uart-has-rtscts` optional property to indicate RTS/CTS hardware flow control and the optional properties  rts-gpios, cts-gpios for RTS/CTS pins. You could check Linux's [Generic Serial DT Bindings](https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/serial/serial.txt). 

Better to add them in `uart.yaml` because most of UART IP blocks support RTS/CTS flow control.

Reply:
@ydamigos i have 2 questions:
1. should i move also tx,rx,cts,rts-pin to uart.yaml ?
2. should i rename `cts/rts-pin` to be like linux `cts/rts-gpios` ?

Reply:
> should i move also tx,rx,cts,rts-pin to uart.yaml ?

I would suggest only cts and rts.

> should i rename cts/rts-pin to be like linux cts/rts-gpios ?

Yes, we try to keep zephyr's DT as close as possible to linux's DT.